File: day-backend/app/presentation/schemas/booking.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field

from app.domain.booking.value_objects import (
    BookingAuditAction,
    BookingSource,
    BookingStatus,
    DepositStatus,
    PaymentMethod,
    PaymentStatus,
    PaymentType,
    RentalMode,
)
from app.domain.property.value_objects import PropertyStatus

# ---------- Guest ----------


class GuestCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    phone: str | None = None
    email: str | None = None
    notes: str | None = None


class GuestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    name: str
    phone: str | None = None
    email: str | None = None
    notes: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class GuestListResponse(BaseModel):
    items: list[GuestResponse]
    total: int
    offset: int
    limit: int


# ---------- Booking ----------


class BookingCreate(BaseModel):
    property_id: uuid.UUID
    guest_name: str = Field(..., min_length=1, max_length=255)
    guest_phone: str | None = None
    guest_email: str | None = None
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode = RentalMode.DAILY
    source: BookingSource = BookingSource.DIRECT
    adults_count: int = Field(default=1, ge=1)
    children_count: int = Field(default=0, ge=0)
    gantt_color: str = "#3B82F6"
    notes: str | None = None


class BookingUpdate(BaseModel):
    check_in: datetime | None = None
    check_out: datetime | None = None
    rental_mode: RentalMode | None = None
    source: BookingSource | None = None
    gantt_color: str | None = None
    gantt_icon: str | None = None
    adults_count: int | None = None
    children_count: int | None = None
    total_price: Decimal | None = None
    notes: str | None = None


class BookingResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    id: uuid.UUID
    company_id: uuid.UUID
    property_id: uuid.UUID
    guest_id: uuid.UUID
    group_booking_id: uuid.UUID | None = None
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode = RentalMode.DAILY
    source: BookingSource
    status: BookingStatus
    gantt_color: str
    gantt_icon: str | None = None
    total_price: Decimal
    calculated_price: Decimal
    # What has actually been received, so a list can show who still owes without
    # asking for every booking's payments one at a time.
    paid_amount: Decimal = Decimal("0")
    adults_count: int
    children_count: int
    notes: str | None = None
    guest_name: str | None = None
    guest_phone: str | None = None
    property_name: str | None = None
    property_internal_name: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class BookingStatusChange(BaseModel):
    # Accept both `target_status` (current) and `status` (legacy tests/clients)
    target_status: BookingStatus = Field(
        validation_alias=AliasChoices("target_status", "status")
    )


class BookingMove(BaseModel):
    target_property_id: uuid.UUID


class BookingListResponse(BaseModel):
    items: list[BookingResponse]
    total: int
    page: int
    per_page: int
    pages: int


# ---------- Payment ----------


class PaymentCreate(BaseModel):
    amount: Decimal = Field(..., gt=0)
    type: PaymentType = PaymentType.PAYMENT
    method: PaymentMethod = PaymentMethod.CASH
    note: str | None = None


class PaymentResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    id: uuid.UUID
    booking_id: uuid.UUID
    amount: Decimal
    type: PaymentType
    method: PaymentMethod
    status: PaymentStatus
    note: str | None = None
    paid_at: datetime | None = None
    created_at: datetime | None = None


# ---------- Deposit ----------


class DepositCreate(BaseModel):
    amount: Decimal = Field(..., gt=0)


class DepositAction(BaseModel):
    action: str = Field(..., pattern="^(pay|return|hold|partial_hold)$")
    held_amount: Decimal | None = None
    reason: str | None = None


class DepositResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    id: uuid.UUID
    booking_id: uuid.UUID
    amount: Decimal
    status: DepositStatus
    held_amount: Decimal
    reason: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


# ---------- File ----------


class BookingFileCreate(BaseModel):
    file_url: str = Field(..., min_length=1, max_length=1024)
    file_name: str = Field(..., min_length=1, max_length=255)
    file_type: str = Field(..., min_length=1, max_length=100)


class BookingFileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    file_url: str
    file_name: str
    file_type: str
    created_at: datetime | None = None


# ---------- Comment ----------


class BookingCommentCreate(BaseModel):
    content: str = Field(..., min_length=1)


class BookingCommentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    author_id: uuid.UUID | None = None
    content: str
    created_at: datetime | None = None


# ---------- Price Calculator ----------


class PriceCalculateRequest(BaseModel):
    property_id: uuid.UUID
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode = RentalMode.DAILY
    adults_count: int = Field(default=1, ge=1)
    children_count: int = Field(default=0, ge=0)


class AvailablePropertyResponse(BaseModel):
    model_config = ConfigDict(json_encoders={Decimal: float})

    property_id: uuid.UUID
    name: str
    internal_name: str
    rental_mode: RentalMode
    rooms: int | None = None
    beds: int | None = None
    total_price: Decimal | None = None
    # Set when the unit is free but has no pricing configured, so the caller can
    # tell "costs nothing" apart from "cannot be quoted".
    price_error: str | None = None


class AvailabilityResponse(BaseModel):
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode
    items: list[AvailablePropertyResponse]


class PriceCalculateResponse(BaseModel):
    model_config = ConfigDict(json_encoders={Decimal: float})

    nights: int
    hours: int
    unit_label: str
    base_total: Decimal
    weekend_surcharge: Decimal
    seasonal_adjustment: Decimal
    extra_guest_surcharge: Decimal
    discount_amount: Decimal
    total: Decimal


# ---------- Gantt ----------


class GanttBookingResponse(BaseModel):
    id: uuid.UUID
    guest_name: str
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode = RentalMode.DAILY
    status: str
    source: str
    gantt_color: str
    gantt_icon: str | None = None
    adults_count: int
    children_count: int
    total_price: float


class GanttPropertyResponse(BaseModel):
    id: uuid.UUID
    name: str
    internal_name: str
    type: str
    status: PropertyStatus
    bookings: list[GanttBookingResponse]


class GanttDataResponse(BaseModel):
    properties: list[GanttPropertyResponse]


# ---------- Today ----------


class TodayBookingItem(BaseModel):
    id: uuid.UUID
    guest_name: str
    property_name: str
    property_internal_name: str
    property_status: PropertyStatus | None = None
    check_in: datetime
    check_out: datetime
    rental_mode: RentalMode = RentalMode.DAILY
    status: BookingStatus
    adults_count: int
    children_count: int


class TodayCheckResponse(BaseModel):
    check_ins: list[TodayBookingItem]
    check_outs: list[TodayBookingItem]
    in_house: list[TodayBookingItem]


# ---------- Audit ----------


class BookingAuditLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    changed_by: uuid.UUID | None = None
    field_name: str
    old_value: str | None = None
    new_value: str | None = None
    action: BookingAuditAction
    created_at: datetime | None = None


# ---------- Full detail ----------


class BookingDetailResponse(BaseModel):
    booking: BookingResponse
    guest: GuestResponse | None = None
    payments: list[PaymentResponse]
    deposits: list[DepositResponse]
    files: list[BookingFileResponse]
    comments: list[BookingCommentResponse]
    contract: None = None  # BookingContractResponse placeholder
    audit_logs: list[BookingAuditLogResponse]

File: day-frontend/src/types/booking.ts

export type BookingStatus = 'pending' | 'confirmed' | 'checked_in' | 'checked_out' | 'completed' | 'cancelled'
export type BookingSource = 'direct' | 'booking' | 'airbnb' | 'other'
export type PaymentType = 'payment' | 'refund'
export type PaymentMethod = 'cash' | 'card' | 'transfer'
export type PaymentStatus = 'pending' | 'completed' | 'failed'
export type DepositStatus = 'pending' | 'paid' | 'returned' | 'held' | 'partially_held'
export type DepositAction = 'pay' | 'return' | 'hold' | 'partial_hold'
export type ContractStatus = 'draft' | 'generated' | 'sent' | 'signed'

export interface Guest {
  id: string
  company_id: string
  name: string
  phone: string
  email: string
  notes: string
  created_at: string
  updated_at: string
}

export interface Booking {
  id: string
  company_id: string
  property_id: string
  guest_id: string
  group_booking_id?: string
  check_in: string
  check_out: string
  source: BookingSource
  status: BookingStatus
  gantt_color: string
  gantt_icon?: string
  total_price: number
  calculated_price: number
  adults_count: number
  children_count: number
  guest_name: string
  guest_phone?: string | null
  property_name: string
  property_internal_name: string
  property_status?: 'new' | 'active' | 'paused' | 'archived'
  created_at: string
  updated_at: string
}

export interface BookingPayment {
  id: string
  booking_id: string
  amount: number
  type: PaymentType
  method: PaymentMethod
  status: PaymentStatus
  note: string
  paid_at: string
  created_at: string
}

export interface BookingDeposit {
  id: string
  booking_id: string
  amount: number
  status: DepositStatus
  held_amount: number
  reason: string
  created_at: string
  updated_at: string
}

export interface BookingFile {
  id: string
  booking_id: string
  file_url: string
  file_name: string
  file_type: string
  created_at: string
}

export interface BookingComment {
  id: string
  booking_id: string
  author_id: string
  content: string
  created_at: string
}

export interface BookingContract {
  id: string
  booking_id: string
  template_url: string
  generated_url: string
  status: ContractStatus
  signed_at: string | null
  created_at: string
}

export interface BookingAuditLog {
  id: string
  booking_id: string
  user_id: string
  action: string
  field_name: string | null
  old_value: string | null
  new_value: string | null
  changed_by: string
  created_at: string
}

export interface BookingDetail {
  booking: Booking
  guest: Guest
  payments: BookingPayment[]
  deposits: BookingDeposit[]
  files: BookingFile[]
  comments: BookingComment[]
  contract?: BookingContract
  audit_logs: BookingAuditLog[]
}

export interface PriceCalculation {
  nights: number
  base_total: number
  weekend_surcharge: number
  seasonal_adjustment: number
  extra_guest_surcharge: number
  discount_amount: number
  total: number
}

export interface GanttPropertySummary {
  id: string
  name: string
  internal_name: string
  type: string
}

export interface GanttData {
  properties: GanttPropertyRow[]
}

export interface GanttPropertyRow extends GanttPropertySummary {
  bookings: Booking[]
}

export interface TodayChecks {
  check_ins: Booking[]
  check_outs: Booking[]
  in_house: Booking[]
}

export interface BookingCreateInput {
  property_id: string
  guest_name: string
  guest_phone: string
  guest_email?: string
  check_in: string
  check_out: string
  source: BookingSource
  adults_count: number
  children_count: number
  gantt_color?: string
  notes?: string
}

export interface BookingUpdateInput {
  check_in?: string
  check_out?: string
  source?: BookingSource
  adults_count?: number
  children_count?: number
  gantt_color?: string
  notes?: string
}

export interface BookingFilters {
  page?: number
  per_page?: number
  status?: BookingStatus
  property_id?: string
  source?: BookingSource
  search?: string
  date_from?: string
  date_to?: string
}

export interface PaymentInput {
  amount: number
  type: PaymentType
  method: PaymentMethod
  note?: string
}

export interface DepositInput {
  amount: number
}

export interface DepositActionInput {
  action: DepositAction
  held_amount?: number
  reason?: string
}

export interface CommentInput {
  content: string
}

export interface PriceCalculateInput {
  property_id: string
  check_in: string
  check_out: string
  adults_count: number
  children_count: number
}

export interface MoveBookingInput {
  target_property_id: string
}

export interface GuestFilters {
  search?: string
  offset?: number
  limit?: number
}

export interface GuestListResponse {
  items: Guest[]
  total: number
  offset: number
  limit: number
}
